### ISSUE TITLE
Export view as image doesn't paint every elements correctly (bmp, png)

### ISSUE THREAD

**Issue by micwic** [archimatetool/archi#439]
### Version of Archi, Operating System

com.archimatetool.editor.buildid=201901281101
org.eclipse.swt.internal.deviceZoom=200

os.name=Windows 10
os.version=10.0
osgi.arch=x86_64

### Expected Behaviour
Export a model view as an image.

### Actual Behaviour
Export a model view as an image. All the elements are not equally and correcty painted.
### Steps to Reproduce the Behaviour
1. Export a model view as an image (first uploaded png file)
2. Copy as Image to Clipboard (second uploaded png file)

[Electro Choc.zip](https://github.com/archimatetool/archi/files/2837389/Electro.Choc.zip)

![electrasmart - product](https://user-images.githubusercontent.com/6366517/52358248-f005f180-2a37-11e9-8453-2d81a1d890c8.png)

![image](https://user-images.githubusercontent.com/6366517/52358430-3eb38b80-2a38-11e9-9bc9-07970a7f7bbd.png)

**Comment by Phillipus** [comment]
I'm assuming you have a hi-res display.

Can you check whether this happens in the previous version of Archi? I made some changes to the resolution of exported images on hi-res systems. I checked on a Retina MacBook and all seemed OK, unless Windows is different.

@jbsarrodie Can you check this JB on hi-res?

**Comment by Phillipus** [comment]
OK, to reproduce this I set the scaling on my Windows machine to 200% and I can the problem, as well as a smaller magic connector cursor and other things.

What I don't understand is why this all works OK on a Retina Mac and not on a Windows hi-res setup?

**Comment by Phillipus** [comment]
In order to fix this problem I'm going to need some help from the users.

On a Retina Mac, the device zoom is reported as 200% but images are 2x too big (exported images, the magic cursor, the format painter cursor and the sticky palette icons in the Sketch and Canvas views) so we render these at 100%

On Windows hi-res, the device zoom is reported as 200% but images are too small unless rendered at 200%.

So what's going on? Anyone like to help?

**Comment by jbsarrodie** [comment]
>Can you check this JB on hi-res?
>So what's going on? Anyone like to help?

Sorry but I don't have access to hi-res monitor (I use full-hd only) :-(

**Comment by Phillipus** [comment]
OK. This is going to be a problem. On my Retina Mac the resolution is 1440x900 so it's not really "hi res" (doubles pixels). But Eclipse SWT reports device zoom as 200. This device zoom number is used to calculate image scale. So on WIndows with a resolution of 2880x1800 device zoom is also 200. Seems no way to know if it's HDPI or not.

**Comment by Phillipus** [comment]
Seems that Modelio tool has the same problem.

**Comment by jbsarrodie** [comment]
>On my Retina Mac the resolution is 1440x900

The physical resolution should be double than that (2 880 × 1 800), thus eclipse reporting device zoom as 200 is normal.

>Seems no way to know if it's HDPI or not.

That's maybe because HDPI doesn't really mean anything...

See https://medium.com/elementaryos/what-is-hidpi-and-why-does-it-matter-b024eabea20d and https://medium.com/elementaryos/top-3-misconceptions-about-hidpi-f5ef493d7bf8

I theory I should be able to force pixel doubling on my FullHD Linux laptop to simulate HDPI, but I still don't know how to do it.

**Comment by Phillipus** [comment]
OK, so the problem is this:

In order to render an image we do this:

**Comment by jbsarrodie** [comment]
Question: why doing this scaling at all ? Isn't it easier to simply export at 100% and let users display this image the way they want ?

My issue here is that as figures don't have real border width, exporting a view at 200% is not equivalent to double pixels of a 100% export as in both cases borders will be only 1px.

**Comment by Phillipus** [comment]
No, because if the user sets the zoom to 200% on Windows exporting at 100% causes the problem.

So, the issue is not about the resolution but the zoom level set in Windows (and Linux?) Mac reports the zoom level as 200% when it's not really 200%

**Comment by jbsarrodie** [comment]
>Mac reports the zoom level as 200% when it's not really 200%

If you have a retina display it **is** 200%. Why do you say it's not?

**Comment by Phillipus** [comment]
I mean that if you scaled the images by 200% on the Mac they are double sized.

**Comment by Phillipus** [comment]
You can set the scaling to 200% in Windows regardless of your screen resolution (will look a bit big) and then deviceZoom is 200. So it's this setting that's the problem. Mac uses 4 pixels per pixels on Retina display, reports 200%, but it behaves like 100%

**Comment by Phillipus** [comment]
I'll try to make this clearer.

On Windows with resolution 1400x900 and scaling set to 100% the device Zoom is 100%
On Windows with resolution 1400x900 and scaling set to 200% the device Zoom is 200%

On Mac with resolution 1400x900 (can't set the scaling) the device Zoom is 200%

So Mac always reports 200%

**Comment by jbsarrodie** [comment]
> On Windows with resolution 1400x900 and scaling set to 100% the device Zoom is 100%
> On Windows with resolution 1400x900 and scaling set to 200% the device Zoom is 200%
> On Retina Mac with resolution 1400x900 (can't set the scaling) the device Zoom is 200%
> So Retina Mac always reports 200%

Ok, this is clearer... and normal: Mac doesn't allow you to change scaling factor as it is built-in. In addition, the resolution reported is not the physical one but the "virtual one" (taking x2 scaling into account). Windows does it the opposite: it always reports (and lets you set) the physical resolution, but lets you choose a scaling factor.

What is unclear to me is which part of the UI are rendered at the physical vs "virtual" resolution.

FWIW, I've just found how it works on Linux and this is exactly the same approach as Windows: you define the physical resolution (1920 x 1080 in my case, i.e. FullHD) and you set the scaling factor (`gsettings set org.gnome.desktop.interface scaling-factor 2`). So I should be able to test if I know what to test exactly...

**Comment by Phillipus** [comment]
What I need to know is when to set the device zoom when getting Image Data:

Normally this would be:

```
int deviceZoom = ImageFactory.getDeviceZoom();
ImageData imageData = image.getImageData(deviceZoom);
```

On WIndows this works fine and images are OK.

On Mac this is setting the images 2x too big in the UI and when exporting views.

I wonder if ImageFactory.getDeviceZoom(); should be set to return 100 if on Mac?

**Comment by jbsarrodie** [comment]
So I did some tests on Linux  (set the scale factor using two different methods) and never go an issue in the UI nor in image export.

When using `gsettings` getDeviceZoom() return according results (ie. 200% with my test) but when using `xrandr` getDeviceZoom() always return 100%.

**Comment by Phillipus** [comment]
Using latest Archi 4.3.2. on Windows with scale factor 200 you'll see:

- tiny magic connector cursor
- tiny format painter cursor
- corrupt exported images

**Comment by Phillipus** [comment]
I'm wondering if this device zoom scaling thing is just a Windows thing? But what about hi-res external monitors for Mac?

I think I'll add a preference "Use device zoom" with defaults of WIndows=on, Mac/Linux=off. Then if users get problems they can experiment with it on/off.

**Comment by Phillipus** [comment]
(The "Close and Comment" button is too big).

**Comment by jbsarrodie** [comment]
>I think I'll add a preference "Use device zoom" with defaults of WIndows=on, Mac/Linux=off. Then if users get problems they can experiment with it on/off.

Might be a good solution

**Comment by Phillipus** [comment]
If deviceZoom is 200% on Windows then exported images are 2x bigger than when deviceZoom is 100%. Is this expected behaviour?

**Comment by Phillipus** [comment]
> My issue here is that as figures don't have real border width, exporting a view at 200% is not equivalent to double pixels of a 100% export as in both cases borders will be only 1px.

And this is actually the cause of the problem where device scaling > 200%.

It's the lines and borders that are not being drawn properly if the device scaling > 200% and the image is exported at 100% scaling. I experimented and set line width to 1 and exported at 100% scaling with Windows set to 200%. The image was OK.

So exporting images at 200% on Windows is actually a workaround for the problem. If we were using actual line widths it would export OK. But then there are other problems with setting line widths (SVG doesn't render properly, zoom diagram not rendering properly).

**Comment by Phillipus** [comment]
> And this is actually the cause of the problem where device scaling > 200%.

Actually it may not be as simple as using proper line width. Exporting a Zest (Visualiser) image also needs to be scaled up otherwise the connections look bad. But that may be due to Zest not using anti-aliasing on connections.

**Comment by Phillipus** [comment]
I've spent a lot of time investigating this.

When creating image data we have to call `Image.getImageData(zoom)` where `zoom` is reported by the system to be 100, 200 on Retina Mac, or another number on Windows depending on the scaling.

On Retina Mac and Linux (I think) if we use a `zoom` of 100 regardless of the `zoom` reported by the system the exported images look OK.

But not so on Windows, because the image size down-scaling messes up the image. So we have to use the `zoom` level reported by the system which results in larger image files.

I don't know why this should be.

**Comment by Phillipus** [comment]
I've created a "zoom" branch with the changes. This code is a mixture of what we had before I changed things to use 100% zoom in commit bfae973f3ad86075004e7903b2d347baab0b2760 and a check to see if we are on Windows or not and if we are, to use the OS scaling factor.

**Comment by Phillipus** [comment]
I guess on Mac retina the resolution is reported as 1400x900 and zoom as 200 so it's not the same as on Windows where the resolution is reported as 2800x1800 and zoom 200?

**Comment by Phillipus** [comment]
So this works on my test configurations:

Windows with scaling of 100, 200 and 300
Linux Mint with scaling of "Double"
Mac Retina

But...I don't know about Mac with external hi-res monitor.

**Comment by Phillipus** [comment]
But of course this now leads to inconsistencies between platforms with hires systems. On Windows with scaling of 200 or greater images are 2x size, but not on Mac or Linux.

Prior to my changes in bfae973f3ad86075004e7903b2d347baab0b2760 all images were exported at 2x when `zoom` was reported as 200.

**Comment by jbsarrodie** [comment]
>On Windows with scaling of 200 or greater images are 2x size,

IMHO the exact image size is not an issue, but the relative width of figure border is. So can you share image export for the same view, but done on different OS so that I can compare them and check if I see any issue?

**Comment by Phillipus** [comment]
The figure border is always the same 1 pixel in all cases. This is a main problem why we have to use 2x export on Windows because it doesn't downscale to 1x very well.

> So can you share image export for the same view, but done on different OS

With the new system or at 100% or....?

**Comment by jbsarrodie** [comment]
>With the new system or at 100% or....?

Sorry. With the new system and scaling set manually to 200% on windows, Retina on Mac and Linux with "double" scaling..

>The figure border is always the same 1 pixel in all cases. This is a main problem why we have to use 2x export on Windows because it doesn't downscale to 1x very well.

I really have to find a Retina or HiDPI laptop to test myself. I really don't understand what this looks like on windows with scaling set to 200% in Archi itself and when exporting...

**Comment by Phillipus** [comment]
I set scaling to 200% in Windows on a non hi-res screen to test. Everything is massive and it's hard to work with but OK for testing.

**Comment by Phillipus** [comment]
Windows with scaling set to 200%

![windows 200 scale](https://user-images.githubusercontent.com/600504/52437703-76d9cd80-2b0f-11e9-8584-84b0c3636f9e.png)

**Comment by Phillipus** [comment]
Mac Retina

![mac retina](https://user-images.githubusercontent.com/600504/52437869-dafc9180-2b0f-11e9-8b09-e4f66abcfb09.png)

**Comment by jbsarrodie** [comment]
>Windows with scaling set to 200%

Thank you, so this is what I feared, as borders are only 1px but figures are twice the size, this no more looks like a "normal" view in Archi.

Can I ask you a screenshot of the whole Archi under windows with 200% scaling and this view opened (and set to 100% in Archi, which is the default) please ?

The question I'm asing to myself is how does it look like when someone works with Archi: does  figure border using only one physical pixel (thus really tiny) or is it automatically scaled and rendered on 2 physical pixels to simulate a "LowDPI" rendering ?

**Comment by Phillipus** [comment]
Linux Double

![linux double](https://user-images.githubusercontent.com/600504/52439006-7858c500-2b12-11e9-987e-bfb4824672fb.png)

**Comment by Phillipus** [comment]
> Can I ask you a screenshot of the whole Archi under windows with 200% scaling and this view opened (and set to 100% in Archi, which is the default) please ?

OK, this was taken on the MacBook Retina with Windows running in a VM and 2880x1800 and scaling at 200%

![windows archi 200](https://user-images.githubusercontent.com/600504/52440684-bce65f80-2b16-11e9-8db6-bc29c7731512.png)

**Comment by jbsarrodie** [comment]
>OK, this was taken on the MacBook Retina with Windows running in a VM and 2880x1800 and scaling at 200%

Thank you.

>The question I'm asing to myself is how does it look like when someone works with Archi: does figure border using only one physical pixel (thus really tiny) or is it automatically scaled and rendered on 2 physical pixels to simulate a "LowDPI" rendering ?

So answer is border is only one physical pixel, thus tiny. So I think we have an issue but are back to borderwidth hack... Don't think we'll solve it anytime soon (for 4.3.3 I mean)

**Comment by Phillipus** [comment]
>  thus tiny.

Bit it looks the same as when using 100% scaling when working in Archi, you don't really know it's any different.

But there are issues around setting the figure line width to an actual amount in Draw2D figures. The line can look different (1) At different zoom levels in Archi (2) When exporting to PNG/BMP/JPG (3) When exporting to SVG. I never could get it working in all cases.

**Comment by Phillipus** [comment]
But for now, the fix I've created will work in order to get images looking OK again on Windows with high scaling. It fixes the regression, so we need to get the next maintenance build out soonish.

**Comment by jbsarrodie** [comment]
>so we need to get the next maintenance build out soonish.

Yes, just wait for me to fix the html/sql export thing (will do during the week-end)

**Comment by Phillipus** [comment]
OK. Let's aim for next week.

**Comment by Phillipus** [comment]
More information...

exporting images at 100% zoom level on a Retina Mac is the right thing to do. Before I looked at this issue images were exported at 200% on Retina Mac. But the quality of the image is not even as good as exporting with the image size set to 200% in the Image export dialog.

This is exported with device scaling at 200% and export image dialog set to 100%. View the image at actual size and you can see the fonts are blurred:

![scaled 200](https://user-images.githubusercontent.com/600504/52482431-29aa3a00-2ba9-11e9-80ce-2a7e19eaee35.png)


This is exported with device scaling at 100% and export image dialog set to 200%. View the image at actual size and you can see the fonts are not blurred:

![exported 200](https://user-images.githubusercontent.com/600504/52482447-3038b180-2ba9-11e9-9335-ec713c044743.png)

**Comment by Phillipus** [comment]
Fixed in Archi 4.3.3. Feel free to re-open or create another issue if there's still a problem.

**Comment by micwic** [comment]
Eveverything ok after this upgrade. Thanks.Avec mes meilleures salutationsKind regardsMichel WickyOsima Conseils, conseils en organisation et technologies de l'informationArchitecture d'entreprise et solutions, Accompagnement du changement, Direction et conduite de projetRuelle de Chaucey 21807 BlonaySwitzerlandmobile +41 79 414 94 44email michel.wicky@osima.chhttp://www.linkedin.com/company/osima-conseilhttp://ch.linkedin.com/in/michelwickyThis message may contain confidential and proprietary material for the sole use of the intended recipient. Any review or distribution by others is strictly prohibited. If you are not the intended recipient, please contact the sender and delete all copies.-----"Phil Beauvoir" <notifications@github.com> a écrit : -----A : "archimatetool/archi" <archi@noreply.github.com>De : "Phil Beauvoir" <notifications@github.com>Date : 12/02/2019 13:06Cc: "Michel Wicky" <michel.wicky@osima.ch>, "Author" <author@noreply.github.com>Objet : Re: [archimatetool/archi] Export view as image doesn't paint every elements correctly (bmp, png) (#439)Fixed in Archi 4.3.3. Feel free to re-open or create another issue if there's still a problem.
